### PR TITLE
Require 128 bit strength encryption key for LE Security mode 1 Level 4

### PR DESCRIPTION
**Pull request by joerchan**
In Bluetooth 5 the definition of LE security mode 1, has changed.
    LE Security Mode 1 level 4 requires authenticated LE Secure Connections
    pairing with encryption using a 128-bit strength encryption key.
    This also changes the behaviour when a security request and response
    would end up with a security level that is lower than the one requested.
    Before pairing would complete, and the link would disconnect with error
    authentication failure. Instead a SMP will abort pairing with error code
    authentication requirement, or encryption key size.

### CODE REVIEW

File: subsys/bluetooth/host/smp.c [hunk 1]
```diff
@@ -239,10 +239,10 @@ struct bt_smp_br {
 	u8_t			remote_dist;
 
 	/* Encryption Key Size used for connection */
-	u8_t enc_key_size;
+	u8_t 			enc_key_size;
```
Comment (carlescufi) [inline subsys/bluetooth/host/smp.c]: What exactly did you change in this structure? just tabbing?

File: subsys/bluetooth/host/smp.c [hunk 2]
```diff
@@ -2601,10 +2611,18 @@ static u8_t smp_pairing_rsp(struct bt_smp *smp, struct net_buf *buf)
...
+	     smp->method == JUST_WORKS) {
 		return BT_SMP_ERR_AUTH_REQUIREMENTS;
 	}
 
+	if ((IS_ENABLED(CONFIG_BT_SMP_SC_ONLY) ||
+	     conn->required_sec_level == BT_SECURITY_FIPS) &&
+	     get_encryption_key_size(smp) != BT_SMP_MAX_ENC_KEY_SIZE) {
+		return BT_SMP_ERR_ENC_KEY_SIZE;
+	}
+
```
Comment (Vudentz) [inline subsys/bluetooth/host/smp.c]: The checks here and in smp_pairing_req are quite similar, I wonder if wouldn't make sense to make get_pair_method to do these checks directly. Also Id assume required_sec_level should always be BT_SECURITY_FIPS if SC_ONLY is enabled, so perhaps required_sec_level is not being initialized properly.

File: subsys/bluetooth/host/smp.c (hunk 1 above)
Comment (carlescufi) [reply subsys/bluetooth/host/smp.c]: You should not use tabs for this btw, only spaces.

File: subsys/bluetooth/host/smp.c [hunk 3]
```diff
@@ -2448,10 +2449,18 @@ static u8_t smp_pairing_req(struct bt_smp *smp, struct net_buf *buf)
 
 	smp->method = get_pair_method(smp, req->io_capability);
 
-	if (IS_ENABLED(CONFIG_BT_SMP_SC_ONLY) && smp->method == JUST_WORKS) {
+	if ((IS_ENABLED(CONFIG_BT_SMP_SC_ONLY) ||
```
Comment (carlescufi) [inline subsys/bluetooth/host/smp.c]: Shouldn't this be `&&`?

Comment (carlescufi) [reply subsys/bluetooth/host/smp.c]: No, it's fine, missed the extra parents.

**Review by carlescufi — APPROVED** [review]

File: subsys/bluetooth/host/smp.c (hunk 2 above)
Comment (joerchan) [reply subsys/bluetooth/host/smp.c]: "Also Id assume required_sec_level should always be BT_SECURITY_FIPS if SC_ONLY is enabled, so perhaps required_sec_level is not being initialized properly."
required_sec_level is initialized properly.
 I added  || req == FIPS  for the case when SC_ONLY is not enabled, but bt_security(FIPS) was called. I want it to behave in the same way then.

Your other point, the code in req/rsp is quite similar, but there are some differences. How much would you suggest we make common?

**Review by jhedberg — APPROVED** [review]

**Review by sjanc — APPROVED** [review]

**Pull request merged**